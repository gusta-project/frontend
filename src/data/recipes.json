[
  {
    "id": 1,
    "name": "Thanks, Obama",
    "user": "JosefBud",
    "userId": 5420,
    "date": "March 20 2019, 16:39:00",
    "notes": "IMPORTANT: CAP Lemon Lime's gravity (g per ml) by default is way above what it should be. If you mix by weight, mix CAP Lemon Lime as 1g = 1ml.\n \n CAP Super Sweet is optional. \n \n VapeWild doesn't add cooling, but 2 drops per 30ml of WS-23 works well if you're into that.\n \n No steep required, shake and vape!",
    "shakeAndVape": true,
    "steepDays": 0,
    "maxVG": false,
    "percentPG": 30,
    "percentVG": 70,
    "tags": ["refreshing", "fruity", "lemon", "lime", "summer"],
    "version": "2.1",
    "flavors": [
      {
        "vendor": "Flavor West",
        "abbreviation": "FW",
        "name": "Blue Raspberry",
        "id": 51,
        "percent": 3
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Cherry Extract",
        "id": 52,
        "percent": 1.5
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Lemon Lime",
        "id": 53,
        "percent": 4
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Super Sweet",
        "id": 54,
        "percent": 0.5
      }
    ]
  },
  {
    "id": 2,
    "name": "DaMomma's RY4 Delight",
    "user": "Alisa",
    "userId": 1024,
    "date": "May 26 2016, 08:10:00",
    "notes": "I am really enjoying this one. The Catalan Cream just works really well with RY4 too.\n \n Aged only a few days and it's YUMMY.",
    "shakeAndVape": false,
    "steepDays": 4,
    "maxVG": false,
    "percentPG": 30,
    "percentVG": 70,
    "tags": ["tobacco", "RY4", "cream", "custard"],
    "version": "2.1",
    "flavors": [
      {
        "vendor": "FlavourArt",
        "abbreviation": "FA",
        "name": "Catalan Cream",
        "id": 55,
        "percent": 1
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "RY4 Double",
        "id": 56,
        "percent": 8
      },
      {
        "vendor": "Flavor West",
        "abbreviation": "FW",
        "name": "Sweet Cream",
        "id": 57,
        "percent": 2.5
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Vanilla Custard v2",
        "id": 58,
        "percent": 2
      }
    ]
  },
  {
    "id": 3,
    "name": "Strawberry Fog",
    "user": "spectrometer",
    "userId": 5823,
    "date": "August 14 2015, 23:09:00",
    "notes": "",
    "shakeAndVape": true,
    "steepDays": 0,
    "maxVG": true,
    "percentPG": null,
    "percentVG": null,
    "tags": ["dessert", "strawberry", "cream", "custard", "heavy"],
    "version": "2.1",
    "flavors": [
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Bavarian Cream",
        "id": 59,
        "percent": 3
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Cheesecake (Graham Crust)",
        "id": 60,
        "percent": 5
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Strawberry Ripe",
        "id": 61,
        "percent": 10
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Vanilla Custard",
        "id": 62,
        "percent": 2
      }
    ]
  },
  {
    "id": 4,
    "name": "Mother's Unicorn Milk",
    "user": "apwroblewski",
    "userId": 9395,
    "date": "March 20 2015, 07:41:00",
    "notes": "",
    "shakeAndVape": true,
    "steepDays": 0,
    "maxVG": false,
    "percentPG": 50,
    "percentVG": 50,
    "tags": ["dessert", "strawberry", "cream", "custard", "heavy"],
    "version": "2.1",
    "flavors": [
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Bavarian Cream",
        "id": 59,
        "percent": 4
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Cheesecake (Graham Crust)",
        "id": 60,
        "percent": 5
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Strawberry Ripe",
        "id": 61,
        "percent": 10
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Sweet Cream",
        "id": 63,
        "percent": 4
      },
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Vanilla Custard",
        "id": 62,
        "percent": 3
      }
    ]
  },
  {
    "id": 5,
    "name": "Ella's Shortbread Biscuit",
    "user": "VHM",
    "userId": 7582,
    "date": "March 26 2016, 07:38:00",
    "notes": "Comforting crumbly buttery vanilla Scottish shortbread, great as a stand alone or as a base for fruits, creams and all sorts of other flavours.\n \n I enjoy this fresh but if you find INW Biscuit too raw the mix really starts to settle after three days getting richer as it ages.\n \n CAP Sugar Cookie and INW Biscuit - they are great cookie bases that pack a punch. Together they work well at mimicking a proper butter shortbread.\n \n AP brings forward the grain of the mix, adding just the right amount of bite and the CAP V.Custard (v.1) keeps the mix from being too dry on the pallet giving that overall buttery feel.\n \n INW Vanilla Shisha added for its inherent bakery notes and general vanilla decadence.\n \n Feeling lazy? Buy the one shot here: \n UK: https://www.chefsflavours.co.uk/vaping-home-mixers/4160-ellas-shortbread-biscuit.html\n USA: http://www.onestopdiyshop.com/product-p/vhmella30.html\n \n Credit: Betamax (Ella Robinson)",
    "shakeAndVape": false,
    "steepDays": 14,
    "maxVG": false,
    "percentPG": 20,
    "percentVG": 80,
    "tags": ["dessert", "cookie", "biscuit", "vanilla", "light"],
    "version": "2.1",
    "flavors": [
      {
        "vendor": "The Flavor Apprentice",
        "abbreviation": "TFA",
        "name": "Acetyl Pyrazine 5%",
        "id": 64,
        "percent": 0.5
      },
      {
        "vendor": "Inawera",
        "abbreviation": "INW",
        "name": "Biscuit",
        "id": 65,
        "percent": 1.5
      },
      {
        "vendor": "Inawera",
        "abbreviation": "INW",
        "name": "Shisha Vanilla",
        "id": 66,
        "percent": 1
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Sugar Cookie",
        "id": 67,
        "percent": 4
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Vanilla Custard",
        "id": 68,
        "percent": 2
      }
    ]
  },
  {
    "id": 6,
    "name": "Stash Comparator Test",
    "user": "JosefBud",
    "userId": 5420,
    "date": "July 28 2019, 19:14:00",
    "notes": "Sometimes you just gotta test things the hard way",
    "shakeAndVape": true,
    "steepDays": 0,
    "maxVG": false,
    "percentPG": 30,
    "percentVG": 70,
    "tags": ["digusting", "dontmixthis", "justatest"],
    "version": "2.1",
    "flavors": [
      {
        "vendor": "Flavor West",
        "abbreviation": "FW",
        "name": "Blue Raspberry",
        "id": 69,
        "percent": 3
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Vanilla Whipped Cream",
        "id": 1,
        "percent": 8
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Lemon Lime",
        "id": 53,
        "percent": 4
      },
      {
        "vendor": "Capella",
        "abbreviation": "CAP",
        "name": "Boston Cream Pie",
        "id": 20,
        "percent": 0.5
      }
    ]
  }
]
