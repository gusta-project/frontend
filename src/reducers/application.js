import { buildActions } from 'utils';

export const types = buildActions('application', [
  'INIT_APP',
  'LOGIN_USER',
  'REQUEST_TOKEN',
  'RECEIVE_TOKEN',
  'REQUEST_CURRENT_USER',
  'RECEIVE_CURRENT_USER',
  'LOGIN_USER_SUCCESS',
  'LOGIN_USER_FAILURE',
  'LOGOUT_USER',
  'LOGOUT_USER_SUCCESS',
  'LOGOUT_USER_FAILURE',
  'REGISTER_USER',
  'REGISTER_USER_SUCCESS',
  'REGISTER_USER_FAILURE'
]);

const initApp = () => ({
  type: types.INIT_APP
});

const loginUser = (emailAddress, password) => ({
  type: types.LOGIN_USER,
  emailAddress,
  password
});

const requestToken = (emailAddress, password) => ({
  type: types.REQUEST_TOKEN,
  emailAddress,
  password
});

const receiveToken = (token, expiration) => ({
  type: types.RECEIVE_TOKEN,
  expiration,
  token
});

const requestCurrentUser = () => ({
  type: types.REQUEST_CURRENT_USER
});

const receiveCurrentUser = user => ({
  type: types.RECEIVE_CURRENT_USER,
  user
});

const loginUserSuccess = user => ({
  type: types.LOGIN_USER_SUCCESS,
  user
});

const loginUserFailure = error => ({
  type: types.LOGIN_USER_FAILURE,
  error
});

const logoutUser = () => ({
  type: types.LOGOUT_USER
});

const logoutUserSuccess = () => ({
  type: types.LOGOUT_USER_SUCCESS
});

const logoutUserFailure = error => ({
  type: types.LOGOUT_USER_FAILURE,
  error
});

const registerUser = details => ({
  type: types.REGISTER_USER,
  details
});

const registerUserSuccess = () => ({
  type: types.REGISTER_USER_SUCCESS
});

const registerUserFailure = error => ({
  type: types.REGISTER_USER_FAILURE,
  error
});

export const actions = {
  initApp,
  loginUser,
  requestToken,
  receiveToken,
  requestCurrentUser,
  receiveCurrentUser,
  loginUserSuccess,
  loginUserFailure,
  logoutUser,
  logoutUserSuccess,
  logoutUserFailure,
  registerUser,
  registerUserSuccess,
  registerUserFailure
};

export const initialState = {
  loggingIn: false,
  loggingOut: false,
  user: null,
  error: null,
  authorization: {
    accessToken: null,
    refreshToken: null,
    expiration: null
  },
  registration: {
    registering: false,
    complete: false,
    error: null
  }
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.LOGIN_USER:
      return {
        ...state,
        loggingIn: true
      };
    case types.LOGOUT_USER:
      return {
        ...state,
        loggingOut: true
      };
    case types.REGISTER_USER:
      return {
        ...state,
        registration: {
          ...state.registration,
          registering: true,
          details: action.details
        }
      };
    case types.RECEIVE_TOKEN:
      return {
        ...state,
        authorization: {
          accessToken: action.token,
          expiration: action.expiration
        }
      };
    case types.RECEIVE_CURRENT_USER:
      return {
        ...state,
        user: action.user
      };
    case types.LOGOUT_USER_SUCCESS:
      return {
        ...state,
        loggingOut: false,
        user: null,
        authorization: initialState.authorization
      };
    case types.LOGIN_USER_SUCCESS:
      return {
        ...state,
        loggingIn: false
      };
    case types.LOGIN_USER_FAILURE:
    case types.LOGOUT_USER_FAILURE:
      return {
        ...state,
        loggingIn: false,
        loggingOut: false,
        error: action.error
      };
    case types.REGISTER_USER_SUCCESS:
      return {
        ...state,
        registration: {
          registering: false,
          complete: true,
          error: null
        }
      };
    case types.REGISTER_USER_FAILURE:
      return {
        ...state,
        registration: {
          registering: false,
          complete: true,
          error: action.error
        }
      };
    default:
      return state;
  }
};
