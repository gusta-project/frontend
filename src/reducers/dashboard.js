import { buildActions } from 'utils';

export const types = buildActions('dashboard', [
  'REQUEST_DASHBOARD',
  'REQUEST_DASHBOARD_SUCCESS',
  'SELECT_DASHBOARD',
  'SELECT_DASHBOARD_SUCCESS',
  'REQUEST_MIGRATIONS',
  'REQUEST_MIGRATIONS_SUCCESS',
  'REQUEST_MIGRATIONS_FAILURE',
  'REQUEST_STATS',
  'REQUEST_STATS_SUCCESS',
  'REQUEST_STATS_FAILURE'
]);

const requestDashboard = () => ({
  type: types.REQUEST_DASHBOARD
});

const requestDashboardSuccess = (dashboardComponent) => ({
  type: types.REQUEST_DASHBOARD_SUCCESS,
  dashboardComponent
});

const selectDashboard = ({ name, item }) => ({
  type: types.SELECT_DASHBOARD,
  name,
  item
});

const selectDashboardSuccess = (dashboardComponent) => ({
  type: types.SELECT_DASHBOARD_SUCCESS,
  dashboardComponent
});

const requestMigrations = () => ({
  type: types.REQUEST_MIGRATIONS
});

const requestMigrationsSuccess = (migrations) => ({
  type: types.REQUEST_MIGRATIONS_SUCCESS,
  migrations
});

const requestMigrationsFailure = (error) => ({
  type: types.REQUEST_MIGRATIONS_FAILURE,
  error
});

const requestStats = () => ({
  type: types.REQUEST_STATS
});

const requestStatsSuccess = (stats) => ({
  type: types.REQUEST_STATS_SUCCESS,
  stats
});

const requestStatsFailure = (error) => ({
  type: types.REQUEST_STATS_FAILURE,
  error
});

export const actions = {
  requestDashboard,
  requestDashboardSuccess,
  selectDashboard,
  selectDashboardSuccess,
  requestMigrations,
  requestMigrationsSuccess,
  requestMigrationsFailure,
  requestStats,
  requestStatsSuccess,
  requestStatsFailure
};

export const initialState = {
  dashboardComponent: {
    name: 'Home',
    item: null
  },
  migrations: [],
  stats: {}
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.REQUEST_DASHBOARD_SUCCESS:
      return {
        ...state,
        dashboardComponent: action.dashboardComponent
      };
    case types.SELECT_DASHBOARD_SUCCESS:
      return {
        ...state,
        dashboardComponent: action.dashboardComponent
      };
    case types.REQUEST_MIGRATIONS_SUCCESS:
      return {
        ...state,
        migrations: action.migrations
      };
    case types.REQUEST_MIGRATIONS_FAILURE:
      return {
        ...state,
        error: action.error
      };
    case types.REQUEST_STATS_SUCCESS:
      return {
        ...state,
        stats: action.stats
      };
    case types.REQUEST_STATS_FAILURE:
      return {
        ...state,
        error: action.error
      };
    default:
      return state;
  }
};
