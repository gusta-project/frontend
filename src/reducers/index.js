import { combineReducers } from 'redux';

import {
  reducer as application,
  initialState as applicationState
} from './application';
import { reducer as profile, initialState as profileState } from './profile';
import { reducer as toast, initialState as toastState } from './toast';

export const initialState = {
  application: applicationState,
  profile: profileState,
  toast: toastState
};

export default combineReducers({
  application,
  profile,
  toast
});
