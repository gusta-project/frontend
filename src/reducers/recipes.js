import { buildActions } from 'utils';

export const types = buildActions('recipe', [
  'REQUEST_RECIPES',
  'REQUEST_RECIPES_SUCCESS',
  'REQUEST_RECIPES_FAILURE'
]);

const requestRecipes = (id) => ({
  type: types.REQUEST_RECIPES,
  id
});

const requestRecipesSuccess = (recipes) => ({
  type: types.REQUEST_RECIPES_SUCCESS,
  recipes
});

const requestRecipesFailure = (error) => ({
  type: types.REQUEST_RECIPES_FAILURE,
  error
});

export const actions = {
  requestRecipes,
  requestRecipesSuccess,
  requestRecipesFailure
};

export const initialState = {
  list: [],
  error: null
};

export const reducer = (state = initialState, action = {}) => {
  switch (action.type) {
    case types.REQUEST_RECIPES_FAILURE:
      return {
        ...state,
        error: action.error
      };
    case types.REQUEST_RECIPES_SUCCESS:
      return {
        ...state,
        list: action.recipes
      };
    default:
      return state;
  }
};
