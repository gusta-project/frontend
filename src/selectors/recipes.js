import { createSelector } from 'reselect';

export const getRecipes = (state) => state.recipes;

export const getRecipeList = createSelector(
  getRecipes,
  (recipes) => recipes.list
);
