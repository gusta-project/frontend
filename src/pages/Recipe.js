import PropTypes from 'prop-types';
import { Helmet } from 'react-helmet';
import queryString from 'query-string';
import React, { Component } from 'react';
import { actions } from 'reducers/application';
import { bindActionCreators } from 'redux';
import { connect } from 'react-redux';

import {
  Container,
  Row,
  Col,
  Table,
  Button,
  ButtonGroup,
  Card,
  Badge
} from 'react-bootstrap';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';

import flavorStash from '../data/flavorstash.json';
import recipes from '../data/recipes.json';

export class Recipe extends Component {
  static propTypes = {
    appActions: PropTypes.object.isRequired,
    location: PropTypes.object
  };

  constructor(props) {
    super(props);

    this.state = {
      flavors: [],
      favorited: false,
      favoriteIcon: ['far', 'heart'],
      rating: 0
    };

    this.handleFavoriteClick = this.handleFavoriteClick.bind(this);
    this.handleRatingClick = this.handleRatingClick.bind(this);
    this.renderRatingButtons = this.renderRatingButtons.bind(this);
  }

  componentDidMount() {
    this.findRecipe();
    this.compareWithStash();
  }

  handleFavoriteClick() {
    const { appActions } = this.props;

    if (this.state.favorited) {
      this.setState({
        favorited: false,
        favoriteIcon: ['far', 'heart']
      });

      appActions.popToast({
        title: 'Success!',
        icon: null,
        message: 'This recipe has been removed from your favorites'
      });
    } else {
      this.setState({
        favorited: true,
        favoriteIcon: ['fas', 'heart']
      });

      appActions.popToast({
        title: 'Success!',
        icon: null,
        message: 'This recipe has been added your favorites'
      });
    }
  }

  handleRatingClick(ratingNumber) {
    if (ratingNumber === this.state.rating) {
      this.setState({ rating: 0 });
    } else {
      this.setState({
        rating: ratingNumber
      });
    }
  }

  findRecipe() {
    const query = this.props.location?.search;
    const queryValues = queryString.parse(query);
    const pageId = parseFloat(queryValues.id);

    const recipe = recipes.filter(e => e.id === pageId)[0];

    recipe.flavorTotal = 0;
    recipe.flavors.forEach(flavor => {
      recipe.flavorTotal += flavor.percent;
    });

    if (recipe.notes) {
      const newNotes = recipe.notes
        .toString()
        .split('\n')
        .map((line, i) => {
          return <Card.Text key={`${recipe.id}${i}`}>{line}</Card.Text>;
        });

      recipe.notes = newNotes;
    }

    const newTags = recipe.tags.map((tag, i) => {
      return (
        <a key={`${recipe.id}${i}`} href="/">
          <Badge className="link--tags">{tag}</Badge>
        </a>
      );
    });

    recipe.tags = newTags;

    this.setState(recipe);
  }

  compareWithStash() {
    const query = this.props.location?.search;
    const queryValues = queryString.parse(query);
    const pageId = parseFloat(queryValues.id);

    const recipe = recipes?.filter?.(e => e.id === pageId)[0];

    let index = 0;

    recipe.flavors.forEach(recipeFlavor => {
      let stashIndex = 0;

      flavorStash.forEach(stashFlavor => {
        if (recipeFlavor.id === stashFlavor.id) {
          recipe.flavors[index].inStash = true;
          return;
        }

        stashIndex++;
        if (stashIndex === flavorStash.length) {
          recipe.flavors[index].inStash = false;
          return;
        }
      });

      index++;
    });

    this.setState(recipe);
  }

  renderRatingButtons(ratingNum) {
    const buttons = [];

    for (let i = 0; i < 5; i++) {
      buttons.push(
        <Button
          className="rating-button button--recipe"
          key={i}
          onClick={() => this.handleRatingClick(i + 1)}
        >
          <span>
            {i < ratingNum ? (
              <FontAwesomeIcon icon={['fas', 'star']} />
            ) : (
              <FontAwesomeIcon icon={['far', 'star']} />
            )}
          </span>
        </Button>
      );
    }

    return <ButtonGroup>{buttons}</ButtonGroup>;
  }

  render() {
    const {
      favoriteIcon,
      rating,
      name,
      version,
      notes,
      userId,
      user,
      date,
      maxvg,
      pg,
      vg,
      shakeNVape,
      steepTime,
      flavorTotal,
      flavors,
      tags
    } = this.state;

    return (
      <Container className="text-center container--recipe">
        <Helmet title="Recipe" />
        <Row className="justify-content-center">
          <Col xs="auto">
            <Button className="button-animation button--recipe">
              <span>Make this recipe</span>
            </Button>
          </Col>
          <Col xs="auto">
            <Button
              onClick={this.handleFavoriteClick}
              className="button--favorite button--recipe"
            >
              <span>
                <FontAwesomeIcon icon={favoriteIcon} />
              </span>
            </Button>
          </Col>
          <Col xs="auto">{this.renderRatingButtons(rating)}</Col>
        </Row>
        <hr />
        <Row>
          <Col lg={{ span: 8, offset: 2 }} xs={{ span: 12 }}>
            <Card className="text-center">
              <Card.Header>{name}</Card.Header>
              <Card.Body>
                <Row>
                  <Col md={{ span: 5 }} xs={{ span: 12 }}>
                    <img
                      src="/media/card-test-1.jpg"
                      alt="card test"
                      className="w-75"
                    />
                  </Col>
                  <Col md={{ span: 7 }} xs={{ span: 12 }}>
                    <Card.Title>Version {version}</Card.Title>
                    <Card.Title>Notes</Card.Title>
                    {notes ? (
                      <div>{notes}</div>
                    ) : (
                      <div>
                        This user did not enter any notes for this recipe.
                      </div>
                    )}
                  </Col>
                </Row>
              </Card.Body>
              <Card.Footer className="text-muted">
                <span>
                  Created by <a href={'/user?id=' + userId}>{user}</a> on {date}
                </span>
              </Card.Footer>
            </Card>
          </Col>
        </Row>
        <Row>
          <Col lg={{ span: 8, offset: 2 }} xs={{ span: 12 }}>
            <Table striped bordered>
              <caption className="text-center">
                {maxvg ? (
                  <span>Max VG |&nbsp;</span>
                ) : (
                  <span>
                    {pg}% PG / {vg}% VG |&nbsp;
                  </span>
                )}
                {shakeNVape ? (
                  <span>Shake &amp; Vape |&nbsp;</span>
                ) : (
                  <span>Steep for {steepTime} days |&nbsp;</span>
                )}
                <span>Flavor total: {flavorTotal}%</span>
              </caption>
              <thead>
                <tr>
                  <th>Flavor</th>
                  <th>Percent</th>
                  <th>Stash</th>
                </tr>
              </thead>
              <tbody>
                {flavors.map((flavor, index) => (
                  <tr key={index}>
                    <td>
                      <a href={'/flavor?id=' + flavor.id}>
                        {flavor.abbreviation} {flavor.name}
                      </a>
                    </td>
                    <td>{flavor.percent.toFixed(1)}%</td>
                    <td>
                      {flavor.inStash ? (
                        <FontAwesomeIcon icon={['far', 'check-square']} />
                      ) : (
                        <FontAwesomeIcon icon={['far', 'square']} />
                      )}
                    </td>
                  </tr>
                ))}
              </tbody>
            </Table>
          </Col>
        </Row>
        <Row>
          <Col lg={{ span: 6, offset: 3 }} xs={{ span: 12 }}>
            <h3>Tags:</h3>
            <div className="d-flex justify-content-center flex-wrap">
              {tags}
            </div>
          </Col>
        </Row>
      </Container>
    );
  }
}

const mapDispatchToProps = dispatch => ({
  appActions: bindActionCreators({ ...actions }, dispatch)
});

export default connect(
  null,
  mapDispatchToProps
)(Recipe);
