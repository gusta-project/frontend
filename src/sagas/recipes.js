import { call, all, put, takeLatest } from 'redux-saga/effects';

import request from 'utils/request';
import { actions, types } from 'reducers/recipes';

function* requestRecipesWorker({ search }) {
  try {
    const endpoint = {
      url: '/recipes',
      method: 'GET',
      data: search
    };
    const result = yield call(request.execute, { endpoint });

    if (!result.success) {
      throw result.error;
    }

    const { data: recipes } = result.response;

    yield put(actions.requestRecipesSuccess(recipes));
  } catch (error) {
    yield put(actions.requestRecipesFailure(error));
  }
}

function* requestRecipesWatcher() {
  return yield takeLatest(types.REQUEST_RECIPES, requestRecipesWorker);
}

export const watchers = {
  requestRecipesWatcher
};

export const workers = {
  requestRecipesWorker
};

export default function* saga() {
  yield all(Object.values(watchers).map((watcher) => watcher()));
}
