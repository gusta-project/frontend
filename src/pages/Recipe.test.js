import React from 'react';
import { Provider } from 'react-redux';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';

import { initialState } from 'reducers';
import { Recipe } from './Recipe';
import recipes from '../data/recipes.json';

const mockStore = configureStore();
const store = mockStore({ application: initialState });

describe('<Recipe />', () => {
  const actions = {
    registerUser: jest.fn(),
    popToast: jest.fn()
  };
  const props = (search) => ({
    location: {
      search
    }
  });

  it('renders correctly', () => {
    const component = renderer.create(
      <Provider store={store}>
        <Recipe appActions={actions} {...props('?id=1')} />
      </Provider>
    );

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('can findRecipe', () => {
    const component = renderer.create(
      <Recipe appActions={actions} {...props('?id=1')} />
    );
    const { instance } = component.root.findByType(Recipe);

    expect(instance).toBeDefined();
    instance.findRecipe();
    expect(instance.state).toEqual({
      ...recipes[0],
      flavorTotal: 9,
      favoriteIcon: ['far', 'heart'],
      favorited: false,
      rating: 0
    });
  });

  it('handles favorite clicks', () => {
    const component = renderer.create(
      <Recipe appActions={actions} {...props('?id=1')} />
    );
    const { instance } = component.root.findByType(Recipe);

    instance.handleFavoriteClick();
    expect(instance.state.favoriteIcon).toEqual(['fas', 'heart']);
    expect(instance.state.favorited).toEqual(true);

    instance.handleFavoriteClick();
    expect(instance.state.favoriteIcon).toEqual(['far', 'heart']);
    expect(instance.state.favorited).toEqual(false);
  });

  it('handles rating clicks', () => {
    const component = renderer.create(
      <Recipe appActions={actions} {...props('?id=1')} />
    );
    const { instance } = component.root.findByType(Recipe);

    instance.handleRatingClick(2);
    expect(instance.state.rating).toEqual(2);
  });

  it('compares flavors against stash', () => {
    const component = renderer.create(
      <Recipe appActions={actions} {...props('?id=1')} />
    );
    const { instance } = component.root.findByType(Recipe);

    instance.compareWithStash();

    for (let i = 0; i < 4; i++) {
      expect(instance.state.flavors[i].inStash).toEqual(false);
    }
  });
});
