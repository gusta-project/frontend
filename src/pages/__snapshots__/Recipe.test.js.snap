// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Recipe /> renders correctly 1`] = `
<div
  className="text-center container--recipe container"
>
  <div
    className="justify-content-center row"
  >
    <div
      className="col-auto"
    >
      <button
        className="button-animation button--recipe btn btn-primary"
        disabled={false}
        type="button"
      >
        <span>
          Make this recipe
        </span>
      </button>
    </div>
    <div
      className="col-auto"
    >
      <button
        className="button--favorite button--recipe btn btn-primary"
        disabled={false}
        onClick={[Function]}
        type="button"
      >
        <span>
          <i
            className="fa far-heart"
          />
        </span>
      </button>
    </div>
    <div
      className="col-auto"
    >
      <div
        className="btn-group"
        role="group"
      >
        <button
          className="rating-button button--recipe btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            <i
              className="fa far-star"
            />
          </span>
        </button>
        <button
          className="rating-button button--recipe btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            <i
              className="fa far-star"
            />
          </span>
        </button>
        <button
          className="rating-button button--recipe btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            <i
              className="fa far-star"
            />
          </span>
        </button>
        <button
          className="rating-button button--recipe btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            <i
              className="fa far-star"
            />
          </span>
        </button>
        <button
          className="rating-button button--recipe btn btn-primary"
          disabled={false}
          onClick={[Function]}
          type="button"
        >
          <span>
            <i
              className="fa far-star"
            />
          </span>
        </button>
      </div>
    </div>
  </div>
  <hr />
  <div
    className="row"
  >
    <div
      className="col-lg-8 col-12 offset-lg-2"
    >
      <div
        className="text-center card"
      >
        <div
          className="card-header"
        >
          Thanks, Obama
        </div>
        <div
          className="card-body"
        >
          <div
            className="row"
          >
            <div
              className="col-md-5 col-12"
            >
              <img
                alt="card test"
                className="w-75"
                src=""
              />
            </div>
            <div
              className="col-md-7 col-12"
            >
              <div
                className="card-title h5"
              >
                Version 
                2.1
              </div>
              <div
                className="card-title h5"
              >
                Notes
              </div>
              <div>
                <p
                  className="card-text"
                >
                  IMPORTANT: CAP Lemon Lime's gravity (g per ml) by default is way above what it should be. If you mix by weight, mix CAP Lemon Lime as 1g = 1ml.
                </p>
                <p
                  className="card-text"
                >
                   
                </p>
                <p
                  className="card-text"
                >
                   CAP Super Sweet is optional. 
                </p>
                <p
                  className="card-text"
                >
                   
                </p>
                <p
                  className="card-text"
                >
                   VapeWild doesn't add cooling, but 2 drops per 30ml of WS-23 works well if you're into that.
                </p>
                <p
                  className="card-text"
                >
                   
                </p>
                <p
                  className="card-text"
                >
                   No steep required, shake and vape!
                </p>
              </div>
            </div>
          </div>
        </div>
        <div
          className="text-muted card-footer"
        >
          <span>
            Created by 
            <a
              href="/user?id=5420"
            >
              JosefBud
            </a>
             on 
            March 20 2019, 16:39:00
          </span>
        </div>
      </div>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-lg-8 col-12 offset-lg-2"
    >
      <table
        className="table table-striped table-bordered"
      >
        <caption
          className="text-center"
        >
          <span>
            30
            % PG / 
            70
            % VG | 
          </span>
          <span>
            Shake & Vape | 
          </span>
          <span>
            Flavor total: 
            9
            %
          </span>
        </caption>
        <thead>
          <tr>
            <th>
              Flavor
            </th>
            <th>
              Percent
            </th>
            <th>
              Stash
            </th>
          </tr>
        </thead>
        <tbody>
          <tr>
            <td>
              <a
                href="/flavor?id=51"
              >
                FW
                 
                Blue Raspberry
              </a>
            </td>
            <td>
              3.0
              %
            </td>
            <td>
              <i
                className="fa far-square"
              />
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="/flavor?id=52"
              >
                TFA
                 
                Cherry Extract
              </a>
            </td>
            <td>
              1.5
              %
            </td>
            <td>
              <i
                className="fa far-square"
              />
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="/flavor?id=53"
              >
                CAP
                 
                Lemon Lime
              </a>
            </td>
            <td>
              4.0
              %
            </td>
            <td>
              <i
                className="fa far-square"
              />
            </td>
          </tr>
          <tr>
            <td>
              <a
                href="/flavor?id=54"
              >
                CAP
                 
                Super Sweet
              </a>
            </td>
            <td>
              0.5
              %
            </td>
            <td>
              <i
                className="fa far-square"
              />
            </td>
          </tr>
        </tbody>
      </table>
    </div>
  </div>
  <div
    className="row"
  >
    <div
      className="col-lg-6 col-12 offset-lg-3"
    >
      <h3>
        Tags:
      </h3>
      <div
        className="d-flex justify-content-center flex-wrap"
      >
        <a
          href="/"
        >
          <span
            className="link--tags badge"
          >
            refreshing
          </span>
        </a>
        <a
          href="/"
        >
          <span
            className="link--tags badge"
          >
            fruity
          </span>
        </a>
        <a
          href="/"
        >
          <span
            className="link--tags badge"
          >
            lemon
          </span>
        </a>
        <a
          href="/"
        >
          <span
            className="link--tags badge"
          >
            lime
          </span>
        </a>
        <a
          href="/"
        >
          <span
            className="link--tags badge"
          >
            summer
          </span>
        </a>
      </div>
    </div>
  </div>
</div>
`;
